### PR TITLE
Window functions

### PR DESCRIPTION
**Pull request by magnatelee**
This PR contains implementations for the following window functions:

* `cunumeric.barlett`
* `cunumeric.blackman`
* `cunumeric.hamming`
* `cunumeric.hanning`
* `cunumeric.kaiser`

### CODE REVIEW

**Comment by ipdemes** [comment]
@magnatelee what are the rules of including third-party library code like `cephes` inside of the cunumeric repo?  Should we clone it as a submodule (if it has git repo)? manually copying source files doesn't feel like a good idea to me unless this is the only way we can use this library.
If we end-up copying source files manually, should we add license information in the headers?

**Comment by magnatelee** [comment]
@ipdemes Please find the `LICENSES_bundled.txt` file that this PR is adding. We should use standalone packages for third-party libraries whenever they are available. One thing unique about `cephes` is that it predates all this packing stuff and is released only in the form of source code. (google it yourself to find it out.) I followed what NumPy and SciPy have been doing for these "embedded" dependencies.

File: src/cunumeric/nullary/window_omp.cc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+template <WindowOpCode OP_CODE>
+struct WindowImplBody<VariantKind::OMP, OP_CODE> {
+  void operator()(
+    AccessorWO<double, 1> out, const Rect<1>& rect, bool dense, int64_t M, double beta) const
+  {
+    WindowOp<OP_CODE> gen(M, beta);
+#pragma omp parallel for schedule(static)
+    for (int64_t idx = rect.lo[0]; idx <= rect.hi[0]; ++idx) out[idx] = gen(idx);
+  }
```
Comment (ipdemes) [inline src/cunumeric/nullary/window_omp.cc:33]: can we add logic for `dense` case here?

File: src/cunumeric/nullary/window_util.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  {
+    return idx < alpha_ ? idx / alpha_ : 2.0 - idx / alpha_;
+  }
+
+  double alpha_;
+};
+
+template <>
+struct WindowOp<WindowOpCode::BLACKMAN> {
+  WindowOp(int64_t M, double beta) : alpha_(M_PI * 2 / (M - 1)) {}
```
Comment (ipdemes) [inline src/cunumeric/nullary/window_util.h:71]: should it be `alpha_(M_PI*2/M)` instead?

File: src/cunumeric/nullary/window_util.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  constexpr double operator()(int64_t idx) const
+  {
+    using std::cos;
```
Comment (ipdemes) [inline src/cunumeric/nullary/window_util.h:74]: in some places you have `using std::cos;` when in others you call `std::cos` directly (line 85). Should it be consistent?

File: src/cunumeric/nullary/window_util.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  assert(false);
+  return f.template operator()<WindowOpCode::BARLETT>(std::forward<Fnargs>(args)...);
+}
+
+template <WindowOpCode OP_CODE>
+struct WindowOp;
+
+template <>
+struct WindowOp<WindowOpCode::BARLETT> {
+  WindowOp(int64_t M, double beta) : alpha_(static_cast<double>(M - 1) / 2.0) {}
```
Comment (ipdemes) [inline src/cunumeric/nullary/window_util.h:60]: Here and  in other specializations below (except KAISER) `beta` is not used, so the signature can be `WindowOp(int64_t M, double )`

File: src/cunumeric/nullary/window_util.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    auto val    = (idx - alpha_) / alpha_;
+    auto result = cyl_bessel_i0(beta_ * sqrt(1 - val * val));
+    result /= cyl_bessel_i0(beta_);
+    return result;
+  }
+#else
+  double operator()(int64_t idx) const
+  {
+    using std::sqrt;
+    auto val    = (idx - alpha_) / alpha_;
```
Comment (ipdemes) [inline src/cunumeric/nullary/window_util.h:115]: should  val be `val = idx/alpha_`

File: tests/window.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import numpy as np
+
+import cunumeric as num
+
+fns = ["bartlett", "blackman", "hamming", "hanning"]
```
Comment (ipdemes) [inline tests/window.py:21]: how about `kaiser`?

Comment (ipdemes) [reply tests/window.py:21]: Actually, I see it is tested below

**Comment by ipdemes** [comment]
@magnatelee : I think we might want to add tests for the cases when `M=0` and `M<0`

File: src/cunumeric/nullary/window_omp.cc (hunk 1 above)
Comment (magnatelee) [reply src/cunumeric/nullary/window_omp.cc:39]: a dense path is added now

File: src/cunumeric/nullary/window_util.h (hunk 3 above)
Comment (magnatelee) [reply src/cunumeric/nullary/window_util.h:74]: I used it when I had to repeat that function more than once. I made the rest of the code consistent.

File: src/cunumeric/nullary/window_util.h (hunk 5 above)
Comment (magnatelee) [reply src/cunumeric/nullary/window_util.h:113]: The code I have here is what NumPy and CuPy does. `idx / alpha_` is greater than 1 for the second half of the array and makes the subsequent `sqrt` call return a nan.

**Comment by magnatelee** [comment]
> @magnatelee : I think we might want to add tests for the cases when M=0 and M<0
done. I added tests for them.

File: src/cunumeric/nullary/window_util.h (hunk 5 above)
Comment (ipdemes) [reply src/cunumeric/nullary/window_util.h:113]: I see, thanks

File: src/cunumeric/nullary/window_util.h (hunk 2 above)
Comment (ipdemes) [reply src/cunumeric/nullary/window_util.h]: @magnatelee can you comment on this? 
From the numpy documentation, it seems like alpha = M_PI*2/M in this case

**Comment by ipdemes** [comment]
@magnatelee :
I see about the license.

File: src/cunumeric/nullary/window_util.h (hunk 2 above)
Comment (ipdemes) [reply src/cunumeric/nullary/window_util.h]: I see, they probably need to update documentation :)

**Review by ipdemes — APPROVED** [review]

**Pull request merged**